### PR TITLE
WT-13153 Fix data race on btree->rec_max_txn and btree->rec_max_timestamp

### PR DESCRIPTION
**Pull request by luke-pearson**
Additionally create a standard function for setting a maximum of two values atomically.

### CODE REVIEW

File: src/include/misc_inline.h
Comment (tod-johnson-mongodb) [inline src/include/misc_inline.h:333]: This comment is not clear since both #if and #else implementations work.  A better comment would be:
/*
 * This #if is a performance optimization. If *shared_value != current then __atomic_compare_exchange_n() returns the value of *shared_value in current but __wt_atomic_cas64() does not.  Thus __atomic_compare_exchange_n() is more efficient. The #if is needed because Windows does not have __atomic_compare_exchange_n() and must use __wt_atomic_cas64().
 */

**Review by tod-johnson-mongodb — APPROVED** [review]
The code works as written. A comment should be changed.

**Review by ajmorton — APPROVED** [review]
LGTM pending Tod's comment.

As a general non-blocking question do we think setting `btree->rec_max_txn/timestamp` is perf critical? Why not use the normal CAS pattern for here and provide a full fix in WT-11749?

File: src/include/extern.h
Comment (RedBeard0531) [inline src/include/extern.h:2547]: The combination of these two names is unfortunate. If read consistently, it implies that either `maximum64` is a type (clearly not), or that `double` is an operation (which is plausible, atomic `2 * v` or `v << 1` is a reasonable operation). So maybe this should be named something like `__wt_set_max_shared_u64` would be better? That would keep the pattern `__wt_OP_shared_TYPE`. Also by putting `u` in there it gives a clean way to extend to support signed types in the future.

Also I suggest returning the old value of `shared_value` prior to the write. I think a lot of algos will need to know at least if they increased it or were a no-op.

File: src/include/misc_inline.h
Comment (RedBeard0531) [inline src/include/misc_inline.h:328]: If this exits because `!(current < new_value)` you will not get the expected barrier semantics out of this primitive. As much as it pains me to say it, you need to so something like this:

```c
while (true) {
  if (current >= new_val) {
    fetch_add(shared_value, 0); // no-op SeqCst RMW
    break;
  }
  if (compare_exchange(...))
    break;
}
```

That will ensure that this always behaves like a barrier. Note that the `ldumaxal` instruction is defined in such a way that this isn't necesary there, just always use that instruction. Also I'm pretty sure that the x86 `lock; cmpxchg;` instruction also has strong semantics even on failure, so if you want to go to ASM there, you would only need the barrier if you would be bailing before the first cmpxchg (but if you do this we should double check the spec.)

See section 5 of the [proposal](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p0493r5.pdf) to standardize `fetch_max` for a bit of discussion. In a mailing list discussion I tried to argue against this, but I think they convinced me that it is at least observable in the C++ memory model, even if only under very contrived examples. Given that all other RMWs are defined as always performing both the read and write (even when no-ops like `fetch_add(0)`, it seems consistent to use the same definition for min/max. Ditto in the WT model where RMWs are defined as "full barriers".

File: src/reconcile/rec_write.c
Comment (RedBeard0531) [inline src/reconcile/rec_write.c:416]: Is this a correctness bug because it implies that `btree->rec_max_foo` can go backwards? If so, can you make sure this ends up in the queue for 8.0 (and backports). Also, we don't have release semantics here, which at least seems questionable...

**Review by RedBeard0531 — CHANGES_REQUESTED** [review]

File: src/reconcile/rec_write.c
Comment (luke-pearson) [reply src/reconcile/rec_write.c:416]: This is a correctness bug, so I will request a backport to 8.0 assuming the change is zero risk.

File: src/include/misc_inline.h
Comment (luke-pearson) [reply src/include/misc_inline.h:326]: Well that is a fairly niche instruction, cool that it exists. I am concerned that we'd be adding another section of ARM specific assembly. ARM assembly requires testing, could at some point change, and is not clear to the reader although given the context of the function it should be obvious as to what it does. 

At this stage the new function is not being called from anywhere in a performance path, in the future it could be. As it stands adding this assembly feels like a case of early optimization with a decent overhead to support that optimization. For that reason I am declining to add this change unless there is a bigger need for it.

File: src/include/misc_inline.h
Comment (luke-pearson) [reply src/include/misc_inline.h:328]: The two usages we have for this act as a `set-and-forget` model, which I don't think requires release semantics, however I could see usages in the future that may want release semantics. This touches on a not well defined part of WiredTiger where we don't specify what memory ordering semantics a function call gives you. If we were to add this I believe we would mandate a minimum ordering of release and not leave it up to the user to specify, this would then be added to the function comment. 

I will need to think about this more. The document has this to say:
"if the user requested memory order is not a release, then store is not required" to confirm the three memory orders that aren't a release are relaxed, acquire, and consume? Where seq_cst is an upgrade on a release.

cc: @ajmorton

Comment (luke-pearson) [reply src/include/misc_inline.h:328]: So I've been thinking about the release semantic part of this, I can't see why we'd need it at least given the two usages in the code that we have now. I do understand that if we don't need to store anything and don't enter the loop we lose any ordering semantics, so we could strengthen it to release semantics (loads and stores don't cross).

I was discussing this with @dgomezferro, and we even considered a relaxed CAS which seemed to provide sufficient semantics given this [cppmem](http://svr-pes20-cppmem.cl.cam.ac.uk/cppmem/index.html) test:
```
int main() {
  atomic_int x = 2;
  {{{ 
   {
   int y = x.load(mo_relaxed).readsvalue(2);
   y = cas_strong_explicit(x, y, 5, mo_relaxed, mo_relaxed);
}
  ||| 
{
   int z = x.load(mo_relaxed);
   z = cas_strong_explicit(x, z, 6, mo_relaxed, mo_relaxed);
   z.readsvalue(6);
}
  }}};
  x.load().readsvalue(5);
  return 0; }
```

Obviously to be consistent we will use a `SEQ_CST` CAS but it did raise the question of what ordering do we need here. I will continue to think on this and see if I can describe a scenario in WT that would be end badly if we didn't at least have release semantics on call to this function.

**Comment by luke-pearson** [comment]
I have updated this PR to drop the correct semantic CAS and replace it with just the standard WiredTiger mechanism. This may seem like an odd choice but I have done this as there is a ticket to fix this here: [WT-11749](https://jira.mongodb.org/browse/WT-11749) and it would be best to not have custom workarounds in the codebase.

File: src/include/extern.h
Comment (luke-pearson) [reply src/include/extern.h]: [20c04cb](https://github.com/wiredtiger/wiredtiger/pull/10740/commits/20c04cb0c22d6ced07dc4804d9c1d2d97fcb6310) I haven't yet implemented the `shared_value` return, I think sometimes things are better added as needed rather than proactively? I need to fish around the codebase for other places we set shared maximums.

File: src/include/misc_inline.h
Comment (luke-pearson) [inline src/include/misc_inline.h:327]: Do either of loads create an additional data race or does the CAS prevent that? Applying the C standard definition of a "data-race"
> The execution of a program contains a data race if it contains two potentially concurrent conflicting actions, at least one of which is not atomic, and neither happens before the other, except for the special case for signal handlers described below [omitted]. Any such data race results in undefined behavior.

Given the CAS orders the first load before and subsequent loads after I think we are safe.

File: src/include/misc_inline.h
Comment (RedBeard0531) [reply src/include/misc_inline.h:326]: What do you mean by "could at some point change"? It is part of the armv8.1 ISA specification. That isn't something that just changes (except to make stronger guarantees than before).

I certainly won't require you to use asm here, but I do think it is worth it to make your atomic primitives as efficient as possible on your supported platforms (while still providing the desired guarantees). I recognize that the current use case may not be perf critical, but once the primitive is available it may be used in more places.

Comment (luke-pearson) [reply src/include/misc_inline.h:326]: It may be used in more places but the CAS loop is the standard way of implementing it, until we get gcc support for `LDUMAXAL` we'll leave it as it.

File: src/include/misc_inline.h
Comment (luke-pearson) [reply src/include/misc_inline.h:328]: This value is read in two additional places:
bt_sync.c:281
```
                if (mod != NULL && mod->rec_max_txn > btree->rec_max_txn)
                    btree->rec_max_txn = mod->rec_max_txn;
                if (mod != NULL && btree->rec_max_timestamp < mod->rec_max_timestamp)
                    btree->rec_max_timestamp = mod->rec_max_timestamp;
```
bt_sync.c:370:
```
        if (!btree->modified && !F_ISSET(conn, WT_CONN_RECOVERING | WT_CONN_CLOSING_CHECKPOINT) &&
          (btree->rec_max_txn >= txn->snapshot_data.snap_min ||
            (conn->txn_global.checkpoint_timestamp != conn->txn_global.last_ckpt_timestamp &&
              btree->rec_max_timestamp > conn->txn_global.checkpoint_timestamp)))
            __wt_tree_modify_set(session);
```
These are all within the context of the `btree->flush_lock` . Now it is evident that the reconciliation code does not take this lock. So there is still the potential to race here. With `rec_max_txn` an `rec_max_timestamp` being written in that order it would be important to read them in that order which does define a `release` requirement on at least `rec_max_txn` and acquire semantics for `rec_max_timestamp`.  

If we write A->B we should read B->A which the code currently does not do. The easiest way to fix this would be to lock in the reconciliation code but that could lead to deadlocks. So for now I will do the following:
1. Guarantee release semantics for the function call into `__wt_set_shared_maximum_u64`
2. Fix bt_sync.c:281 to utilize the CAS functionality
3. Modify the order of reads and perform an acquire load at bt_sync.c:370

Comment (luke-pearson) [reply src/include/misc_inline.h:328]: I was thinking of message passing when it came to ordering these writes, I still can't define a valid order between them. This would be much easier if there was a lock to utilize here.

Comment (luke-pearson) [reply src/include/misc_inline.h:328]: I don't think there's any way to do this obviously correctly other than using a lock. In theory any key value pair in the tree with txnid or timestamp greater than the btree->rec_max_txn and btree->rec_max_timestamp should trigger the `__wt_tree_modify_set`.

Which means even if we locked around writing to these fields in reconciliation there's nothing that would prevent another write happening which would then move these values. I think without a fully fledged review of this code and the related code paths I won't have a reliable solution here. I am probably going to kick off a google doc.

**Comment by luke-pearson** [comment]
This needs more thought and time spent figuring out what the intended use case is for these variables.

File: src/include/misc_inline.h
Comment (marcbutler) [reply src/include/misc_inline.h:328]: @luke-pearson When I tried the cppmem test posted above in cppm I got: 112 executions; no consistent.
I'm curious as to which model you used (none worked for me) and how you drew your conclusion. Thanks.

Comment (luke-pearson) [reply src/include/misc_inline.h:328]: That is correct there is no consistent execution that will end with reading the value 5 from x at the end of the execution. In thread 1 we attempt to swap 2 with 5, that is the state that `x.readsvalue(2)` creates. We don't check this for success as we don't have another readsvalue check for that thread. Thread 2 attempts to swap in 6 for whatever value exists. Given thread 1 reads 2 we have two scenarios:

1. Thread 1 CAS Success: The value 5 was swapped in but we didn't read that value at the end of execution, this means that 2 was overwritten by 6
2. Thread 1 CAS Failure: Nothing was swapped in by thread 1 as the thread 2 CAS happened after we read 2, therefore 6 must exist at the end.

I'm sure you get the point but yes there's no way to reach an end state which has a value of 5 for x.

Comment (marcbutler) [reply src/include/misc_inline.h:328]: Interesting. I think a constraint is missing, the following gives a consistent execution:
```
int main() {
  atomic_int x = 2;
  {{{ 
   {
   int y = x.load(mo_relaxed); // X
   y = cas_strong_explicit(x, y, 5, mo_relaxed, mo_relaxed);
   y.readsvalue(5);
}
  ||| 
{
   int z = x.load(mo_relaxed).readsvalue(2);
   z = cas_strong_explicit(x, z, 6, mo_relaxed, mo_relaxed);
   z.readsvalue(6);
}
  }}};
  x.load().readsvalue(5);
  return 0; }
```
But adding `readsvalue(2);` at position `X` forces a data race  on x.

Comment (dgomezferro) [reply src/include/misc_inline.h:328]: @marcbutler as far as I know cpp mem doesn't support ifs/whiles, that's why I simulated the execution I was interested in with `.readsvalue()`s. If you remove `readsvalue(2);` from the first thread then it will read 6 and overwrite it with 5, which is something that wouldn't happen with Luke's C implementation (we'd see that 5 < 6 and bail out).

**Pull request closed without merging**
